### DISCUSSION TITLE
Scoped instance (ISession NHibernate) is opening new sessions when using ConstructServicesUsing with DI

### DISCUSSION THREAD

**Discussion by marciliooliveira** [LuckyPennySoftware/AutoMapper#4020] [NA]
Hi @jbogard, would you have any idea how we could bind the session when trying to load a related entity from a custom converter using automapper? I've been facing a problem with that using DI (.net core). Since the session is created at any time when the dependency is necessary (the instance is Scoped), and when I resolve the entity using a custom converter on MyCustomConverter class it's in a different session, so if try to save it in my BaseRepository I getting an error like "[HibernateException: illegally attempted to associate a proxy with two open Sessions]" 

The ISession instance is requested scoped, but when I try to get an instance of some repository from the current provider (ServiceCollection) in order to inject it into the custom converter, I get a null and I have to register it in a new provider scope and it is when the new ISession come up. My repositories are registered as transient.

Something like that.. just simplified for understanding

```
var config = new MapperConfiguration(config =>
            {
                    config.ConstructServicesUsing(type => {

                    var entityType = type.GetGenericArguments()[0];
                    var repositoryType = typeof(InterfaceMyRepository<>).MakeGenericType(entityType);
                    var repository = services.BuildServiceProvider()
                                             .GetRequiredService(repositoryType);  // Always NULL/exception if I don't registered it previously
                                                                                // If I registered it services.AddTransient(repositoryType) 
                                                                                // A new ISession is opened

                    ActivatorUtilities.CreateInstance(provider, type, repository);
            }
});
```

And in my customConverter

```
public class EntityConverter<T> : ITypeConverter<long, T>,
                                      ITypeConverter<long?, T>
            where T : MyEntityBase
    {
        private readonly IMyRepository<T> _myRepository;

        public EntityConverter(IMyRepository<T> myRepository)
        {
            _myRepository = myRepository;
        }

        public T Convert(long? source, T destination, ResolutionContext context)
        {
            //Load the related entity from the source, but in a different session
           return _myRepository.Load(source);
        }
}
```

Thank you!

**Comment by lbargaoanu** [comment]
https://jimmybogard.com/automapper-usage-guidelines/
You seem to have moved the logic of your app in the AM configuration, don't do that :)
You don't need `ConstructServicesUsing`, just use the DI package.

**Reply by jbogard** [reply]
There's also not enough in the question here to understand what's going on in the registration or resolution. But the answer is most definitely NOT to use `ConstructServicesUsing`. The DI package is what you want.

In any case, this is probably a better question for Stack Overflow since it involves a lot more moving pieces than AutoMapper.
